jgrapht-core/src/main/java/org/jgrapht/graph/specifics/FastLookupUndirectedSpecifics.java

```diff
@@ -0,0 +1,146 @@
+ * -----------------
+ * (C) Copyright 2015-2015, by Joris Kinable and Contributors.
+ *
+ * Original Author:  Joris Kinable
+ * Contributor(s):
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ */
+package org.jgrapht.graph.specifics;
+
+import org.jgrapht.Graph;
+import org.jgrapht.graph.AbstractBaseGraph;
+import org.jgrapht.util.ArrayUnenforcedSet;
+import org.jgrapht.util.VertexPair;
+
+import java.io.Serializable;
+import java.util.*;
+
+/**
+ * Fast implementation of UndirectedSpecifics. This class uses additional data structures to improve the performance of methods which depend
+ * on edge retrievals, e.g. getEdge(V u, V v), containsEdge(V u, V v),addEdge(V u, V v). A disadvantage is an increase in memory consumption.
+ * If memory utilization is an issue, use a {@link DirectedSpecifics} instead.
+ *
+ * @author Joris Kinable
+ */
+public class FastLookupUndirectedSpecifics<V,E>
+    extends UndirectedSpecifics<V,E>
+    implements Serializable
+{
+    private static final long serialVersionUID = 225772727571597846L;
+
+    /* Maps a pair of vertices <u,v> to an set of edges {(u,v)}. In case of a multigraph, all edges which touch both u,v are included in the set */
+    protected Map<VertexPair<V>, ArrayUnenforcedSet<E>> touchingVerticesToEdgeMap;
+
+    public FastLookupUndirectedSpecifics(AbstractBaseGraph<V, E> abstractBaseGraph)
+    {
+        this(abstractBaseGraph, new LinkedHashMap<>());
+    }
+
+    public FastLookupUndirectedSpecifics(AbstractBaseGraph<V, E> abstractBaseGraph, Map<V, UndirectedEdgeContainer<V, E>> vertexMap)
+    {
+        super(abstractBaseGraph, vertexMap);
+        this.touchingVerticesToEdgeMap=new HashMap<>();
+    }
+
+    /**
+     * @see Graph#getAllEdges(Object, Object)
+     */
+    @Override public Set<E> getAllEdges(V sourceVertex, V targetVertex)
+    {
+        if (abstractBaseGraph.containsVertex(sourceVertex)&& abstractBaseGraph.containsVertex(targetVertex)) {
+            Set<E> edges = touchingVerticesToEdgeMap.get(new VertexPair<>(sourceVertex, targetVertex));
+            return edges == null ? Collections.emptySet() : new ArrayUnenforcedSet<>(edges);
+        }else{
+            return null;
+        }
+    }
+
+    /**
+     * @see Graph#getEdge(Object, Object)
+     */
+    @Override public E getEdge(V sourceVertex, V targetVertex)
+    {
+        List<E> edges = touchingVerticesToEdgeMap.get(new VertexPair<>(sourceVertex, targetVertex));
+        if(edges==null || edges.isEmpty())
+            return null;
+        else
+            return edges.get(0);
+    }
+
+    @Override public void addEdgeToTouchingVertices(E e)
+    {
+        V source = abstractBaseGraph.getEdgeSource(e);
+        V target = abstractBaseGraph.getEdgeTarget(e);
+
+        getEdgeContainer(source).addEdge(e);
+
+
+        //Add edge to both touchingVerticesToEdgeMap.get(u,v) and touchingVerticesToEdgeMap.get(v,u)
+        VertexPair<V> vertexPair1=new VertexPair<>(source, target);
```

Comment:
We could reduce the memory load by using an UndirectedVertexPair implementation whose equals/hashCode respects the undirected comparison semantics.  That way we can avoid storing the second pair.